### PR TITLE
Issue 199

### PR DESCRIPTION
**Pull request by drumwhizz**

### CODE REVIEW

File: ublox-bluetooth-commands-mode.go [hunk 1]
```diff
@@ -58,13 +59,27 @@ func (ub *UbloxBluetooth) ResetUblox() error {
 // ResetUblox calls the Serial port's ResetViaDTR and does not return until
 // the ublox module has indicated it is ready
 func (ub *UbloxBluetooth) ResetUbloxSync() error {
+	ub.rebootExpected = true
```
Comment (RobHumphris) [inline ublox-bluetooth-commands-mode.go:67]: Should this be protected by a mutex? As there is a wait in the function we might get unexpected flag states.

File: ublox-bluetooth-commands-mode.go [hunk 2]
```diff
@@ -58,13 +59,27 @@ func (ub *UbloxBluetooth) ResetUblox() error {
...
 	if err != nil {
 		return errors.Wrap(err, "[ResetUblox] error")
 	}
 
-	_, err = ub.WaitForResponse(rebootResponseString, false)
+	select {
+	case <-ub.rebootDetected:
 
-	return err
+	case <-time.After(time.Second * 4):
```
Comment (RobHumphris) [inline ublox-bluetooth-commands-mode.go]: Its been a while but I thought we'd found that time.After can be problematic and to use a context with timeout:
```
d := time.Now().Add(shortDuration)
	ctx, cancel := context.WithDeadline(context.Background(), d)

	// Even though ctx will be expired, it is good practice to call its
	// cancellation function in any case. Failure to do so may keep the
	// context and its parent alive longer than necessary.
	defer cancel()

	select {
	case <-time.After(1 * time.Second):
		fmt.Println("overslept")
	case <-ctx.Done():
		fmt.Println(ctx.Err())
	}
```

File: ublox-bluetooth-commands.go [hunk 3]
```diff
@@ -147,36 +155,40 @@ func (btd *BluetoothDevices) MultiDiscoverWithContext(ctx context.Context, scant
 	// Don't concatenate errors. We need to spot the ContextCancelled error
 	err = <-errChan
 	if noOfDevices == 2 {
-		err = <-errChan
+		err2 := <-errChan
```
Comment (RobHumphris) [inline ublox-bluetooth-commands.go:158]: could this cause a block in this thread?

Comment (drumwhizz) [reply ublox-bluetooth-commands.go:158]: If we have two dongles, we will get 2 channels signalled

Comment (RobHumphris) [reply ublox-bluetooth-commands.go:158]: True but the thread depends on both returning, there's always the possiblity of a timeout

File: ublox-bluetooth-commands-mode.go (hunk 1 above)
Comment (drumwhizz) [reply ublox-bluetooth-commands-mode.go:67]: If this was a hard real time application then yes. I'd rather have the function return the error, than block on a mutex(). Easier to spot a problem

File: ublox-bluetooth-commands-mode.go (hunk 2 above)
Comment (drumwhizz) [reply ublox-bluetooth-commands-mode.go]: time.After() can cause memory leaks by not being Garbaged collected...

Comment (RobHumphris) [reply ublox-bluetooth-commands-mode.go]: Confused - we have time.After in the original code, thats why I suggest case <-ctx.Done():

File: ublox-bluetooth-commands.go (hunk 3 above)
Comment (drumwhizz) [reply ublox-bluetooth-commands.go:158]: Code has been in there since dual dongle began and has seen no issue

Comment (drumwhizz) [reply ublox-bluetooth-commands.go:158]: The go function that signals the error channel ensures it will always happen

**Review by RobHumphris — APPROVED** [review]

**Pull request merged**